beacon_node/eth2-libp2p/src/rpc/codec/ssz_snappy.rs

```diff
@@ -0,0 +1,256 @@
+            Err(e) => Err(e),
+        }
+    }
+}
+
+/* Outbound Codec: Codec for initiating RPC requests */
+pub struct SSZSnappyOutboundCodec<TSpec: EthSpec> {
+    encoder: snap::raw::Encoder,
+    decoder: snap::raw::Decoder,
+    len: Option<usize>,
+    protocol: ProtocolId,
+    phantom: PhantomData<TSpec>,
+}
+
+impl<TSpec: EthSpec> SSZSnappyOutboundCodec<TSpec> {
+    pub fn new(protocol: ProtocolId, max_packet_size: usize) -> Self {
+        // this encoding only applies to ssz_snappy.
+        debug_assert!(protocol.encoding.as_str() == "ssz_snappy");
+
+        SSZSnappyOutboundCodec {
+            encoder: snap::raw::Encoder::new(),
+            decoder: snap::raw::Decoder::new(),
+            protocol,
+            len: None,
+            phantom: PhantomData,
+        }
+    }
+}
+
+// Encoder for outbound streams: Encodes RPC Requests to peers
+impl<TSpec: EthSpec> Encoder for SSZSnappyOutboundCodec<TSpec> {
+    type Item = RPCRequest<TSpec>;
+    type Error = RPCError;
+
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let bytes = match item {
+            RPCRequest::Status(req) => req.as_ssz_bytes(),
+            RPCRequest::Goodbye(req) => req.as_ssz_bytes(),
+            RPCRequest::BlocksByRange(req) => req.as_ssz_bytes(),
+            RPCRequest::BlocksByRoot(req) => req.block_roots.as_ssz_bytes(),
+            RPCRequest::Testing(req) => req.as_ssz_bytes(),
+            RPCRequest::Phantom(_) => unreachable!("Never encode phantom data"),
+        };
+        // Compressed RpcRequests are not UVI encoded since they don't have chunks.
+        // Correspondingly on the receiving end, they are directly snappy uncompressed.
+        let compressed_bytes = self.encoder.compress_vec(&bytes).map_err(RPCError::from)?;
+        dst.extend_from_slice(&compressed_bytes);
+        Ok(())
+    }
+}
+
+// Decoder for outbound streams: Decodes RPC responses from peers.
+//
+// The majority of the decoding has now been pushed upstream due to the changing specification.
+// We prefer to decode blocks and attestations with extra knowledge about the chain to perform
+// faster verification checks before decoding entire blocks/attestations.
+impl<TSpec: EthSpec> Decoder for SSZSnappyOutboundCodec<TSpec> {
+    type Item = RPCResponse<TSpec>;
+    type Error = RPCError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // Decode the length of the uncompressed bytes
+        let length = self.len.unwrap_or_else(|| {
+            let (length, remaining) = decode::u64(src).unwrap();
+            let input_len = src.len();
+            let remaining_len = remaining.len();
+            src.split_to(input_len - remaining_len);
+            self.len = Some(length as usize);
+            length as usize
+        });
+
+        let mut reader = FrameDecoder::new(Cursor::new(&src));
+        let mut decoded_buffer = vec![0; length];
```

Comment:
Here we are allocating a buffer of an arbitrary length given to us by a potentially malicious peer. Someone could send a very large number and crash lighthouse nodes. 

We should check the length to be less than a specific size